File: _archive/go-backend/internal/protondb/snapshots.go
package protondb

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"regexp"
	"sort"
	"strings"
	"time"
)

const (
	ReportsContentsAPI = "https://api.github.com/repos/bdefore/protondb-data/contents/reports"
	ReportsRawBaseURL  = "https://raw.githubusercontent.com/bdefore/protondb-data/master/reports"
)

var snapshotNamePattern = regexp.MustCompile(`(?i)^reports_([a-z]{3})(\d{1,2})_(\d{4})\.tar\.gz$`)

var monthByAbbrev = map[string]time.Month{
	"jan": time.January,
	"feb": time.February,
	"mar": time.March,
	"apr": time.April,
	"may": time.May,
	"jun": time.June,
	"jul": time.July,
	"aug": time.August,
	"sep": time.September,
	"oct": time.October,
	"nov": time.November,
	"dec": time.December,
}

// Snapshot describes one reports_*.tar.gz archive in protondb-data.
type Snapshot struct {
	Filename string    `json:"filename"`
	Date     time.Time `json:"date"`
	URL      string    `json:"url,omitempty"`
	Size     int64     `json:"size,omitempty"`
}

// ParseSnapshotFilename extracts the date from names like reports_jun1_2026.tar.gz.
func ParseSnapshotFilename(filename string) (time.Time, bool) {
	matches := snapshotNamePattern.FindStringSubmatch(strings.TrimSpace(filename))
	if matches == nil {
		return time.Time{}, false
	}

	month, ok := monthByAbbrev[strings.ToLower(matches[1])]
	if !ok {
		return time.Time{}, false
	}

	var day, year int
	if _, err := fmt.Sscanf(matches[2], "%d", &day); err != nil {
		return time.Time{}, false
	}
	if _, err := fmt.Sscanf(matches[3], "%d", &year); err != nil {
		return time.Time{}, false
	}

	date := time.Date(year, month, day, 0, 0, 0, 0, time.UTC)
	if date.Day() != day || date.Month() != month || date.Year() != year {
		return time.Time{}, false
	}
	return date, true
}

// SelectLatestSnapshot selects the newest valid reports_*.tar.gz archive name.
func SelectLatestSnapshot(filenames []string) (Snapshot, bool) {
	var latest Snapshot
	found := false
	for _, filename := range filenames {
		date, ok := ParseSnapshotFilename(filename)
		if !ok {
			continue
		}
		if !found || date.After(latest.Date) || (date.Equal(latest.Date) && filename > latest.Filename) {
			latest = Snapshot{
				Filename: filename,
				Date:     date,
				URL:      ReportsRawBaseURL + "/" + filename,
			}
			found = true
		}
	}
	return latest, found
}

// SortSnapshotsNewestFirst sorts snapshots in-place by date descending.
func SortSnapshotsNewestFirst(snapshots []Snapshot) {
	sort.SliceStable(snapshots, func(i, j int) bool {
		if snapshots[i].Date.Equal(snapshots[j].Date) {
			return snapshots[i].Filename > snapshots[j].Filename
		}
		return snapshots[i].Date.After(snapshots[j].Date)
	})
}

type githubContentItem struct {
	Name        string `json:"name"`
	Size        int64  `json:"size"`
	DownloadURL string `json:"download_url"`
	Type        string `json:"type"`
}

// ListSnapshotsFromGitHub lists ProtonDB report archives through the GitHub contents API.
// It only reads directory metadata; it does not download report archives.
func ListSnapshotsFromGitHub(ctx context.Context, client *http.Client) ([]Snapshot, error) {
	if client == nil {
		client = http.DefaultClient
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, ReportsContentsAPI, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "ProtonSage/0.1 (+https://github.com/bdefore/protondb-data)")

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("query ProtonDB reports directory: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return nil, fmt.Errorf("query ProtonDB reports directory: GitHub API returned %s", resp.Status)
	}

	var items []githubContentItem
	if err := json.NewDecoder(resp.Body).Decode(&items); err != nil {
		return nil, fmt.Errorf("decode GitHub contents response: %w", err)
	}

	snapshots := make([]Snapshot, 0, len(items))
	for _, item := range items {
		if item.Type != "file" && item.Type != "" {
			continue
		}
		date, ok := ParseSnapshotFilename(item.Name)
		if !ok {
			continue
		}
		url := item.DownloadURL
		if url == "" {
			url = ReportsRawBaseURL + "/" + item.Name
		}
		snapshots = append(snapshots, Snapshot{
			Filename: item.Name,
			Date:     date,
			URL:      url,
			Size:     item.Size,
		})
	}
	SortSnapshotsNewestFirst(snapshots)
	return snapshots, nil
}

// LatestSnapshotFromGitHub returns the newest ProtonDB report archive metadata.
func LatestSnapshotFromGitHub(ctx context.Context, client *http.Client) (Snapshot, error) {
	snapshots, err := ListSnapshotsFromGitHub(ctx, client)
	if err != nil {
		return Snapshot{}, err
	}
	if len(snapshots) == 0 {
		return Snapshot{}, errors.New("no reports_*.tar.gz snapshots found in ProtonDB reports directory")
	}
	return snapshots[0], nil
}

File: src/protondb/snapshots.cpp
#include "protondb/snapshots.h"
#include <QJsonDocument>
#include <QJsonArray>
#include <QJsonObject>
#include <QRegularExpression>
#include <QNetworkReply>
#include <QUrlQuery>
#include <QEventLoop>
#include <QDebug>
#include <algorithm>

namespace ProtonSage {

// ── Month abbreviation parser ─────────────────────────────────────────

static const QMap<QString, int>& monthAbbrevMap()
{
    static const QMap<QString, int> m = []() {
        QMap<QString, int> map;
        auto add = [&](const char* name, int val) {
            map.insert(QString::fromLatin1(name), val);
        };
        add("jan", 1); add("feb", 2); add("mar", 3);
        add("apr", 4); add("may", 5); add("jun", 6);
        add("jul", 7); add("aug", 8); add("sep", 9);
        add("oct", 10); add("nov", 11); add("dec", 12);
        return map;
    }();
    return m;
}

// ── parseSnapshotFilename ─────────────────────────────────────────────

QDate parseSnapshotFilename(const QString& filename)
{
    static const QRegularExpression re(
        QStringLiteral("^reports_([a-z]{3})(\\d{1,2})_(\\d{4})\\.tar\\.gz$"),
        QRegularExpression::CaseInsensitiveOption);

    const auto match = re.match(filename.trimmed());
    if (!match.hasMatch())
        return {};

    const QString monthStr = match.captured(1).toLower();
    if (!monthAbbrevMap().contains(monthStr))
        return {};

    bool dayOk = false, yearOk = false;
    int day = match.captured(2).toInt(&dayOk);
    int year = match.captured(3).toInt(&yearOk);
    if (!dayOk || !yearOk)
        return {};

    int month = monthAbbrevMap().value(monthStr, 0);
    if (month < 1 || month > 12 || day < 1 || day > 31)
        return {};

    QDate date(year, month, day);
    // Validate that the parsed components survive round-trip through QDate
    // (catches invalid combos like Feb 30).
    if (date.year() != year || date.month() != month || date.day() != day)
        return {};

    return date;
}

// ── selectLatestSnapshot ──────────────────────────────────────────────

Snapshot selectLatestSnapshot(const QStringList& filenames)
{
    Snapshot latest;
    for (const QString& fn : filenames) {
        QDate date = parseSnapshotFilename(fn);
        if (!date.isValid())
            continue;

        bool isNewer = !latest.date.isValid()
                       || date > latest.date
                       || (date == latest.date && fn > latest.filename);
        if (isNewer) {
            latest.filename = fn;
            latest.date = date;
            latest.url = kReportsRawBaseURL + u'/' + fn;
        }
    }
    return latest;
}

// ── sortSnapshotsNewestFirst ──────────────────────────────────────────

void sortSnapshotsNewestFirst(QList<Snapshot>& snapshots)
{
    std::sort(snapshots.begin(), snapshots.end(),
              [](const Snapshot& a, const Snapshot& b) {
                  if (a.date == b.date)
                      return a.filename > b.filename;
                  return a.date > b.date;
              });
}

// ── GitHub API helpers ────────────────────────────────────────────────

static QByteArray syncGet(QNetworkAccessManager& nam, const QUrl& url,
                           QString* error = nullptr)
{
    QNetworkRequest req(url);
    req.setRawHeader("Accept", "application/vnd.github+json");
    req.setRawHeader("User-Agent",
                     "ProtonSage/0.2 (+https://github.com/bdefore/protondb-data)");

    QNetworkReply* reply = nam.get(req);

    // Synchronous wait using a local event loop.
    QEventLoop loop;
    QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
    loop.exec();

    if (reply->error() != QNetworkReply::NoError) {
        if (error)
            *error = QStringLiteral("GitHub API error: %1").arg(reply->errorString());
        reply->deleteLater();
        return {};
    }

    int status = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
    if (status < 200 || status > 299) {
        if (error)
            *error = QStringLiteral("GitHub API returned HTTP %1").arg(status);
        reply->deleteLater();
        return {};
    }

    QByteArray body = reply->readAll();
    reply->deleteLater();
    return body;
}

// ── listSnapshotsFromGitHub ────────────────────────────────────────────

QList<Snapshot> listSnapshotsFromGitHub(QNetworkAccessManager& nam)
{
    QString err;
    QByteArray body = syncGet(nam, QUrl(kReportsContentsAPI), &err);
    if (body.isEmpty()) {
        qWarning() << "listSnapshotsFromGitHub:" << err;
        return {};
    }

    QJsonParseError parseErr;
    QJsonDocument doc = QJsonDocument::fromJson(body, &parseErr);
    if (doc.isNull() || !doc.isArray()) {
        qWarning() << "listSnapshotsFromGitHub: JSON parse error —"
                   << parseErr.errorString();
        return {};
    }

    QList<Snapshot> snapshots;
    QJsonArray items = doc.array();
    for (const QJsonValue& val : items) {
        QJsonObject obj = val.toObject();

        // Only process files (and entries without explicit type).
        QString type = obj[u"type"].toString();
        if (!type.isEmpty() && type != u"file")
            continue;

        QString name = obj[u"name"].toString();
        QDate date = parseSnapshotFilename(name);
        if (!date.isValid())
            continue;

        Snapshot snap;
        snap.filename = name;
        snap.date = date;
        snap.size = static_cast<qint64>(obj[u"size"].toDouble());

        // Prefer download_url; fall back to raw.githubusercontent.com.
        QString dw = obj[u"download_url"].toString();
        if (!dw.isEmpty())
            snap.url = dw;
        else
            snap.url = kReportsRawBaseURL + u'/' + name;

        snapshots.append(snap);
    }

    sortSnapshotsNewestFirst(snapshots);
    return snapshots;
}

// ── latestSnapshotFromGitHub ──────────────────────────────────────────

Snapshot latestSnapshotFromGitHub(QNetworkAccessManager& nam)
{
    QList<Snapshot> snapshots = listSnapshotsFromGitHub(nam);
    if (snapshots.isEmpty()) {
        qWarning() << "latestSnapshotFromGitHub: no snapshots found";
        return {};
    }
    return snapshots.first();
}

} // namespace ProtonSage
